### PR TITLE
Engine discard methods

### PR DESCRIPTION
**Pull request by nolwn**

### CODE REVIEW

File: engine/engine.go [hunk 1]
```diff
@@ -95,6 +95,38 @@ func (game *State) PlayUp(p uint8, c Card) error {
 	return nil
 }
 
+// Discard deletes a card from the hand of the player p
+func (game *State) Discard(p uint8, c Card) error {
```
Comment (nolwn) [inline engine/engine.go]: If we want to use single letter names as indeces, we should probably call `c` `card`. I did the same thing (and forgot to fix it before I merged).

File: engine/engine.go [hunk 2]
```diff
@@ -95,6 +95,38 @@ func (game *State) PlayUp(p uint8, c Card) error {
...
+	pl := &game.players[p]
+	i, err := findCard(pl.hand, c)
+
+	if err != nil {
+		return errors.New("engine.Game.Discard: card not found")
```
Comment (nolwn) [inline engine/engine.go:104]: I like this, especially since there is no stack trace!

File: engine/engine.go [hunk 3]
```diff
@@ -95,6 +95,38 @@ func (game *State) PlayUp(p uint8, c Card) error {
...
+}
+
+// DiscardPlayed played deletes all cards from the game that have been played,
+// and returns a slice of the cards deleted in this way
+func (game *State) DiscardPlayed() (stack []Card) {
+	for p := range game.players {
+		pl := &game.players[p]
+		for _, card := range pl.hand {
+			if card.played == true {
+				game.Discard(uint8(p), card)
```
Comment (nolwn) [inline engine/engine.go]: Not a big deal, but you are double searching here. You already have found the card before you send it to Discard to find again and then remove. Why not just remove it here?

We could make the actual delete logic a helper function if we don't want to repeat the code.

File: engine/engine_test.go [hunk 4]
```diff
@@ -218,11 +218,188 @@ func TestViewAs(t *testing.T) {
 
 	actual := view.players[playerFour].hand[0]
 
+
```
Comment (nolwn) [inline engine/engine_test.go]: Was this extra line on purpose?

File: engine/engine_test.go [hunk 5]
```diff
@@ -218,11 +218,188 @@ func TestViewAs(t *testing.T) {
...
+	game.SetPlayer(0, 0, maxHandSize, cloneHand(handOne))
+
+	spadesAce := Card{true, true, Spades, 0}
+	err := game.Discard(0, spadesAce)
+	if err != nil {
+		t.Error("expected no error, but returned one")
+	}
+
+	expected := []Card{
```
Comment (nolwn) [inline engine/engine_test.go:237]: I like really these; I think it's a good idea not to try to get clever in unit tests and just write out the expected inputs and outputs! Bueno.

File: engine/engine_test.go [hunk 6]
```diff
@@ -218,11 +218,188 @@ func TestViewAs(t *testing.T) {
...
+
+	if expectedCardCount != actualCardCount {
+		t.Errorf("expected cardCount of %d, actual: %d", expectedCardCount, actualCardCount)
+	}
+
+	heartsEight := Card{true, true, Hearts, 8}
+
+	err = game.Discard(0, heartsEight)
+	if err != nil {
+		t.Error("expected no error, but returned one")
```
Comment (nolwn) [inline engine/engine_test.go:269]: It might be more helpful to give some context to this error.

**Comment by nolwn** [comment]
I added an issue for this #6!

**Comment by nolanee** [comment]
resolves #6

**Pull request merged**